fix: key and value typing of unions of object types.

grigorischristainas opened the pull request:
Closes #727.

I thought it was an inference issue so I tried to resolve it by modifying the types and using `infer`. Moreover, I tried to keep using the internal types `EnumerableStringKeyOf` and `EnumerableStringKeyedValueOf`. 

I would be happy to integrate any further adjustments as needed. 

---

Make sure that you:

- [ ] Typedoc added for new methods and updated for changed
- [x] Tests added for new methods and updated for changed
- [ ] New methods added to `src/index.ts`
- [ ] New methods added to `docs/src/pages/mapping.md`

---

<details><summary>We use semantic PR titles to automate the release process!</summary>

https://conventionalcommits.org

PRs should be titled following using the format: `< TYPE >(< scope >)?: description`

### Available Types:

- `feat`: new functions, and changes to a function's type that would impact users.
- `fix`: changes to the runtime behavior of an existing function, or refinements to it's type that shouldn't impact most users.
- `perf`: changes to function implementations that improve a functions _runtime_ performance.
- `refactor`: changes to function implementations that are neither `fix` nor `perf`
- `test`: tests-only changes (transparent to users of the function).
- `docs`: changes to the documentation of a function **or the documentation site**.
- `build`, `ci`, `style`, `chore`, and `revert`: are only relevant for the internals of the library.

For scope put the name of the function you are working on (either new or
existing).

</details>

cjquines reviewed (changes requested):
thanks for the pr!

i suspect we have a similar issue with other places where we use `EnumerableStringKeyedValueOf`, which i think are `pickBy`, `omitBy`, `mapValues`, `forEachObj`, and `values`. for example

```ts
  test("union of records", () => {
    forEachObj(
      {} as Record<number, string> | Record<string, number>,
      (value, key) => {
        expectTypeOf(key).toEqualTypeOf<string | `${number}`>();
        expectTypeOf(value).toEqualTypeOf<number | string>();
      },
    );
  });
```

i suspect the correct fix would have to be at the `EnumerableStringKeyOf` and `EnumearableStringKeyedValueOf` level?

grigorischristainas commented:
Hi @cjquines, I have committed the changes and added some tests. Could you please review them and let me know if any further adjustments are necessary?

cjquines commented on src/internal/types.ts:
```diff
@@ -96,15 +97,25 @@ type IsBoundedString<T> = T extends string
+/**
+ * Extracts the value type from an object type T.
+ */
+export type InferredRecordValue<T> = T extends EmptyObject
+  ? T[keyof T]
```
**question:** why is there a `T extends EmptyObject` check here? what goes wrong without it?

can you add a comment for this

grigorischristainas replied:
This was added mainly due to a failing test case for `values`: 
```ts
it("should correctly type functions", () => {
    const result = values(doNothing());
    expectTypeOf(result).toEqualTypeOf<Array<never>>();
});
```
Without the check for `EmptyObject`, result is typed as `uknown[]` instead of `never[]`. I agree that it is a bit confusing and a comment would be useful. If you have a better suggestion, I would be happy to integrate it; otherwise, I will proceed with adding a comment.

cjquines replied:
makes sense, let's comment

cjquines commented on src/mapValues.test.ts:
```diff
@@ -116,7 +116,7 @@ describe("typing", () => {
       mapValues(userValues, (value, key) => {
         expectTypeOf(value).toEqualTypeOf<number>();
-        expectTypeOf(key).toEqualTypeOf<`${UserID}`>();
+        expectTypeOf(key).toMatchTypeOf<`${UserID}`>();
```
**question:** why doesn't `toEqualTypeOf` work here?

i think units should always assert the strongest possible assertion, and in this case i can't see how `key` can be narrower than `${UserID}`

grigorischristainas replied:
I got an error when using `toEqualTypeOf`, which I suspect has to do with type inference (?). I got it to work with `assertType` but it is not being used in codebase, so I changed it with `toMatchTypeOf`. Have you encountered a similar issue before ?

cjquines replied:
strange indeed… i'm fine with this for now

eranhirsch replied:
Let's add a TODO so we can test this out separately without delaying this PR

cjquines commented on src/omitBy.ts:
```diff
@@ -19,15 +19,19 @@ type PickSymbolKeys<T extends object> = {
-    ReconstructedRecord<T>
-  >
->;
+type PartialEnumerableKeys<T extends object> = T extends unknown
```
**nit:** whenever we have to do the `T extends unknown` trick, we should add the distributive conditional comment, i think? see definition for `IsBoundedKey`

grigorischristainas replied:
Sure, I will add the distributive conditional comment, it will be useful for future reference!

cjquines commented on src/internal/types.ts:
```diff
@@ -96,15 +97,25 @@ type IsBoundedString<T> = T extends string
  * @see EnumerableStringKeyedValueOf
  */
-export type EnumerableStringKeyOf<T> = `${Exclude<keyof T, symbol>}`;
+export type EnumerableStringKeyOf<T> =
+  T extends Record<infer K, unknown> ? `${Exclude<K, symbol>}` : never;
```
**suggestion:** let's add tests for this and `EnumerableStringKeyedValueOf` to `types.test.ts`

testing would be wide cases (`Record<string, string>`, `Record<string | number, string>`, etc.), literals (`{ a: 1 }`, etc.), union types (like we have here), things with `symbol` (which is the intent behind this type in the first place), things with optional properties

grigorischristainas replied:
I agree, this would be very useful. I will add them in `types.test.ts` within the next few days.

cjquines reviewed (dismissed):
this looks awesome, thanks for the thoroughness here! i've got one suggestion to add more tests to `types.test.ts`, but i'm fine with shipping this as-is if you don't have time to add some.

eranhirsch commented on src/internal/types.ts:
```diff
@@ -96,15 +97,25 @@ type IsBoundedString<T> = T extends string
+/**
+ * Extracts the value type from an object type T.
+ */
+export type InferredRecordValue<T> = T extends EmptyObject
```
The "Inferred" part in the name does not describe what this type is; it's how it's calculated. This doesn't communicate clearly what this type is. Similarly, Records, conceptually, are a subset of all objects for simple mappings. It's true that when doing complex typing, it's a nice shorthand for more complicated syntax, but it's not the intent of the term.

Naming stuff is often a matter of taste, so I'm not putting my foot down on this, but having read a lot of old code to try and detangle Remeda in the first place, a good name goes a long way to simplify long-term maintenance.

It seems like the correct name here would be `Value` or `ValuesOf`, as in `ValuesOf<MyType>`

grigorischristainas replied:
Thank you for your suggestion, `ValuesOf` is indeed a more precise name.

eranhirsch commented on src/keys.test.ts:
```diff
@@ -111,6 +112,25 @@ describe("typing", () => {
 
+    test("unions of records", () => {
+      const dataFirst = keys(
+        {} as Record<number, string> | Record<string, string>,
```
You should use `unknown` for the values type, to make it even more general.
```suggestion
        {} as Record<number, unknown> | Record<string, unknown>,
```

eranhirsch commented on src/keys.test.ts:
```diff
@@ -111,6 +112,25 @@ describe("typing", () => {
+      >();
+
+      const dataLast = pipe(
+        {} as Record<number, string> | Record<string, string>,
```
You can pull out this as a data variable and reuse it in the 2 tests instead of repeating the type. That way, if someone changes the type for one test, they don't drift accidentally.

```suggestion
        data,
```

grigorischristainas replied:
I agree it would be better to extract a common `data` variable. I will integrate this change to the rest of the tests too, as you suggested.

eranhirsch commented on src/forEachObj.test.ts:
```diff
@@ -70,4 +70,22 @@ describe("typing", () => {
+    forEachObj(
+      {} as Record<number, string> | Record<string, number>,
+      (value, key) => {
+        expectTypeOf(key).toEqualTypeOf<string>();
```
wouldn't the key here need to be:
```ts
`${number}` | string
```

Is that being squished by TypeScript because they are both strings?

grigorischristainas replied:
I think it is indeed being squished by TypeScript as the first one is a subset of string. I tried this example below:
```ts
type A = `${number}`
type B = string
type C = A | B // C is of type string
```

eranhirsch commented on src/forEachObj.test.ts:
```diff
@@ -70,4 +70,22 @@ describe("typing", () => {
+    );
+
+    pipe(
+      {} as Record<number, string> | Record<string, number>,
```
Similar comments as above, keep a single "data" variable with the type so that both tests are synced and we don't risk changing one without changing the other.
```suggestion
      data,
```

eranhirsch commented on src/keys.ts:
```diff
@@ -50,7 +50,9 @@ type IndicesAfterSpread<
-  T extends Record<PropertyKey, never> ? [] : Array<EnumerableStringKeyOf<T>>;
+  T extends Record<PropertyKey, never>
+    ? []
+    : Array<EnumerableStringKeyOf<Required<T>>>;
```
Shouldn't the `Required` wrapper part be part of the `EnumerableStringKeyOf` type and not here?

grigorischristainas replied:
You are right thank you, I will integrate it in `EnumerableStringKeyOf`.

eranhirsch commented on src/mapKeys.test.ts:
```diff
@@ -105,4 +105,20 @@ describe("typing", () => {
+
+  test("union of records", () => {
+    const dataFirst = mapKeys(
+      {} as Record<number, string> | Record<string, string>,
```
1. similar to previous comments, extract a common `data` 
variable.
2. We don't care about the value type, we should use `unknown` instead of `string` to keep the test generic.
```suggestion
      data,
```

eranhirsch commented on src/mapValues.test.ts:
```diff
@@ -160,4 +160,24 @@ describe("typing", () => {
+
+  test("unions of records", () => {
+    const dataFirst = mapValues(
+      {} as Record<number, string> | Record<string, string>,
```
similar to previous comments: 1. pull a `data` variable out, and use `unknown` in places we don't care about a specific type, like the values here.

eranhirsch commented on src/omitBy.ts:
```diff
@@ -19,15 +19,19 @@ type PickSymbolKeys<T extends object> = {
+        ReconstructedRecord<T>
+      >
+    >
+  : unknown;
```
Shouldn't this be `never`?

eranhirsch commented on src/values.test.ts:
```diff
@@ -53,6 +53,13 @@ describe("typing", () => {
   });
 
+  it("should correctly type union of Records", () => {
+    const result = values<Record<number, string> | Record<string, string>>({
+      a: "b",
+    });
```
Don't use an explicit generic of `values`, define the type for the data itself.
```suggestion
    const result = values({
      a: "b",
    } as Record<number, string> | Record<string, string>);
```

eranhirsch reviewed (commented):
I just have a few nits about styling and test ergonomics.

grigorischristainas commented:
@cjquines @eranhirsch thank you both for your feedback! I have integrated the requested changes and additionally added tests for internal types `EnumerableStringKeyOf` and `EnumerableStringKeyedValueOf`. I would be happy to integrate any additional adjustments if needed.

eranhirsch commented on src/internal/types.ts:
```diff
@@ -8,11 +8,12 @@ import {
 declare const __brand: unique symbol;
 
-export type Branded<T, Brand extends symbol> = T & { [__brand]: Brand };
+export type Branded<T, Brand extends PropertyKey> = T & { [__brand]: Brand };
```
Why the change? We want to limit the usage of branded types as much as possible, but this opens up the usage to regular strings...

grigorischristainas replied:
I see, I changed that to also support a string, as in the related test in `mapValues`. Maybe I could just change the type in the test from:
```ts
type UserID = Branded<string, "UserId">;
```
to
```ts
type UserID = Branded<string, symbol>;
```
and revert that change in `Branded` type.

eranhirsch commented on src/internal/types.ts:
```diff
@@ -96,15 +97,30 @@ type IsBoundedString<T> = T extends string
+/**
+ * Extracts the value type from an object type T.
+ */
+export type ValuesOf<T> =
```
This isn't used outside this file so it shouldn't be exported. We also usually put non-exported "private" types below the exported ones so it's easier to read through the file.

eranhirsch commented on src/mapKeys.test.ts:
```diff
@@ -105,4 +105,14 @@ describe("typing", () => {
     expectTypeOf(result).toEqualTypeOf<Partial<Record<123, string>>>();
   });
+
+  test("union of records", () => {
+    const data = {} as Record<number, string> | Record<string, string>;
+
+    const dataFirst = mapKeys(data, constant("hello" as string));
+    expectTypeOf(dataFirst).toEqualTypeOf<Record<string, string>>();
+
+    const dataLast = pipe(data, mapKeys(constant("hello" as string)));
+    expectTypeOf(dataLast).toEqualTypeOf<Record<string, string>>();
+  });
```
I don't think this test tests what this PR fixes. Because you use constant for the new key, there's nothing testing the result of the union. You should test different value types (and the key doesn't matter): `Record<PropertyKey, "cat"> | Record<PropertyKey, "dog">`, or test the mapper itself like you test in forEach.

eranhirsch commented on src/mapValues.test.ts:
```diff
@@ -106,7 +107,6 @@ describe("typing", () => {
 
   describe("branded types", () => {
     test("should infer types correctly in the mapper", () => {
-      type Branded<K, T> = K & { _type: T };
```
👍🏼

eranhirsch commented on src/values.test.ts:
```diff
@@ -53,6 +53,14 @@ describe("typing", () => {
   });
 
+  it("should correctly type union of Records", () => {
+    const result = values({
+      a: "b",
+    } as Record<number, string> | Record<string, string>);
+
+    expectTypeOf(result).toEqualTypeOf<Array<string>>();
```
This doesn't test the main issue this PR deals with clearly enough; the records should have different value types so that we can see that the result is the correct union of them; using string for both just skips the complexity, and it doesn't matter what the key is.
```suggestion
    const result = values({
      a: "b",
    } as Record<PropertyKey, "cat"> | Record<PropertyKey, "dog">);

    expectTypeOf(result).toEqualTypeOf<Array<string>>();
```

eranhirsch commented on src/internal/types.test.ts:
```diff
@@ -156,3 +161,144 @@ describe("IfBoundedRecord", () => {
+
+  test("union of records", () => {
+    expectTypeOf<
+      EnumerableStringKeyOf<Record<number, unknown> | Record<string, unknown>>
```
Instead of testing number and string, which cause typescript to squish the result type, you can use a template strings which won't be squished with the number string:

```suggestion
      EnumerableStringKeyOf<Record<number, unknown> | Record<`prefix_${string}`, unknown>>
```

eranhirsch commented on src/internal/types.test.ts:
```diff
@@ -156,3 +161,144 @@ describe("IfBoundedRecord", () => {
+
+  test("union keys", () => {
+    expectTypeOf<
+      EnumerableStringKeyOf<Record<number | string, unknown>>
```
similarly.

eranhirsch commented on src/internal/types.test.ts:
```diff
@@ -156,3 +161,144 @@ describe("IfBoundedRecord", () => {
+describe("EnumerableStringKeyedValueOf", () => {
+  test("string values", () => {
+    expectTypeOf<
+      EnumerableStringKeyedValueOf<Record<number, string>>
```
when you don't care about the key use PropertyKey and not number/string/etc...
```suggestion
      EnumerableStringKeyedValueOf<Record<PropertyKey, string>>
```

eranhirsch commented on src/internal/types.test.ts:
```diff
@@ -156,3 +161,144 @@ describe("IfBoundedRecord", () => {
+  test("union of records", () => {
+    expectTypeOf<
+      EnumerableStringKeyedValueOf<
+        Record<number, string> | Record<string, string>
```
Similar to my comment in one of the specific tests, this doesn't test the union logic very well because both records use the same value type. Instead, we should test two disjoint value types.
```suggestion
        Record<number, "cat"> | Record<string, "dog">
```

eranhirsch commented on src/internal/types.test.ts:
```diff
@@ -156,3 +161,144 @@ describe("IfBoundedRecord", () => {
+  });
+
+  test("empty object", () => {
+    const emptyObj = {} as const;
```
Doesn't type-fest already provide a type for this?

grigorischristainas replied:
Are you referring to the `EmptyObject` type? Sure, I could use that one, but it is also used in the `ValuesOf` type. So I thought I might utilize it this way, as I've seen it used in other tests. I could replace it with the `EmptyObject`.

eranhirsch reviewed (changes requested):
Some tests don't test the core of the issue very well. I left some comments to address that.

grigorischristainas commented:
Thank you @eranhirsch, I rebased and integrated the changes for types tests and the `Branded` type in the previous commits [eab3445](https://github.com/remeda/remeda/pull/728/commits/eab344508aaf421b1c097e40eccc26a3d84b4111) & [4ccfddf](https://github.com/remeda/remeda/pull/728/commits/4ccfddfb798de1b236b6bcda4957afb6dadfcd4c). I added the rest of the changes in new commits. Please let me know if any additional adjustments are needed.

eranhirsch reviewed (approved):
This is really good work! thanks for doing it

TkDodo commented:
:tada: This PR is included in version 2.0.11 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/remeda/v/2.0.11)
- [GitHub release](https://github.com/remeda/remeda/releases/tag/v2.0.11)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
